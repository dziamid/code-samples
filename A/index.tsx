/**
 * A link to a certain page, an anchor tag
 */

import * as React from 'react';
import {omit} from 'lodash';
const {createStyleSheet} = require('jss-theme-reactor');
import {Action, ActionCreator, bindActionCreators} from 'redux';
const {connect} = require('react-redux');
import {push, PushAction} from 'react-router-redux';
const autoBind = require('react-autobind');
const cx = require('classnames');

export interface IAProps extends React.ClassAttributes<A> {
  href?: any;
  push?: PushAction;
  children?: React.ReactNode;
  className?: string;
  target?: string;
  title?: string;
  footer?: boolean;
  linkClass?: string;
  disabled?: boolean;
}

export const styleSheet = createStyleSheet('A', () => {
  return {
    root: {
      display: 'inline-block',
      textAlign: 'center',
      textDecoration: 'none',
      color: '#00C5C1',
      verticalAlign: 'middle',
      cursor: 'pointer',
      '&:hover': {
        fontWeight: 700,
      },
      '&:active': {
        color: '#107976',
        fontWeight: '400',
      },
      '&::after': {
        display: 'block',
        content: 'attr(title)',
        fontWeight: 'bold',
        height: '1px',
        color: 'transparent',
        overflow: 'hidden',
        visibility: 'hidden',
      },
    },
    footer: {
      composes: '$root',
      color: 'rgba(117, 117, 117, 0.6)',
      '&:active': {
        color: '#757575',
      },
    },
    disabled: {
      composes: '$root',
      pointerEvents: 'none',
      cursor: 'default',
      '&:hover': {
        fontWeight: 400,
      },
    },
  };
});

export class A extends React.Component<IAProps, {}> {
  static contextTypes = {
    styleManager: React.PropTypes.object.isRequired,
  };

  static defaultProps = {
    footer: false,
    disabled: false,
  };

  constructor(props: IAProps, context: any) {
    super(props, context);

    autoBind(this);
  }

  handleClick(event: React.SyntheticEvent<any>) {
    event.preventDefault();
    this.props.push(this.props.href);
  }

  render() {
    const {className, children, href, target, footer, disabled} = this.props;
    const classes = this.context.styleManager.render(styleSheet);
    const other = omit(this.props, ['className', 'children', 'href', 'target', 'push', 'footer', 'linkClass']);

    const rootClasses = {
      [`${classes.root}`]: true,
      [`${classes.footer}`]: footer,
      [`${classes.disabled}`]: disabled,
    };

    return (
      <a
        href={href}
        onClick={this.handleClick}
        target={target}
        className={this.props.linkClass || cx(rootClasses)}
        {...other}
      >
        { children }
      </a>
    );
  }
}

const mapDispatchToProps = (dispatch: any) => bindActionCreators({
  push,
}, dispatch);

export default connect(null, mapDispatchToProps)(A);
