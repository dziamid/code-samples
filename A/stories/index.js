import React from 'react';
import { storiesOf } from '@kadira/storybook';
import A from '../index';


storiesOf('A', module)
  .add('default', () => (
    <p>
      Lorem ipsum dolor sit amet <A title="consectetur">consectetur</A> adipiscing elit.
    </p>
  ))
  .add('footer color', () => (
    <p>
      Lorem ipsum dolor sit amet <A footer title="consectetur">consectetur</A> adipiscing elit.
    </p>
  ));
