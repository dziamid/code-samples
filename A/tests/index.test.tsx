/**
 * Testing our link component
 */
import * as React from 'react';
import A from '../index';
import {mount} from 'enzyme';
import {default as context} from 'tests/context/all';
const toJson = require('enzyme-to-json').default;

describe('<A />', () => {
  it('should render correctly', () => {
    const children = (<h1>Test</h1>);
    const wrapper = mount(context(
      <A href="/link">
        {children}
      </A>
    ));

    expect(toJson(wrapper.find('A'))).toMatchSnapshot();
  });

});
