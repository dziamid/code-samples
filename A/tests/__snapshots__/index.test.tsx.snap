exports[`<A /> should render correctly 1`] = `
<A
  disabled={false}
  footer={false}
  href="/link"
  push={[Function]}>
  <a
    className="A-root-mock"
    disabled={false}
    href="/link"
    onClick={[Function]}>
    <h1>
      Test
    </h1>
  </a>
</A>
`;
